File: Leetcode/648/648.cpp
class Solution {
public:
    string join(const vector<string>& vec) {
        string str;
        if (!vec.empty()) {
            str += vec[0];
            for (size_t i = 1; i != vec.size(); ++i) {
                str.push_back(' ');
                str += vec[i];
            }
        }
        return str;
    }
    string replaceWords(vector<string>& dictionary, string sentence) {
        unordered_set<string_view> dict(dictionary.begin(),dictionary.end());
        vector<string> vec;
        istringstream iss(sentence);
        string temp;
        do {
            iss >> temp;
            vec.emplace_back(temp);
        } while (iss);
        vec.pop_back();
        for (auto &s : vec) {
            int length = (int)s.size();
            for (int i = 0; i < length; ++i) {
                string_view sv(s.data(), i+1);
                if (dict.find(sv) != dict.end()) {
                    while ((int)s.size() > i+1) {
                        s.pop_back();
                    }
                    break;
                }
            }
        }
        return join(vec);
    }
};

File: Leetcode/648/648.go
func replaceWords(dictionary []string, sentence string) string {
	dict := make(map[string]struct{})
	temp := struct{}{}
	for i := 0; i < len(dictionary); i++ {
		dict[dictionary[i]] = temp
	}
	split := strings.Split(sentence, " ")
	for i := 0; i < len(split); i++ {
		var str []uint8
		for j := 0; j < len(split[i]); j++ {
			str = append(str, split[i][j])
			_, ok := dict[string(str)]
			if ok {
				split[i] = string(str)
				break
			}
		}
	}
	return strings.Join(split, " ")
}